app/split_shares.go

```diff
@@ -0,0 +1,298 @@
+)
+
+// SplitShares uses the provided block data to create a flattened data square.
+// Any MsgWirePayForDatas are malleated, and their corresponding
+// MsgPayForData and Message are written atomically. If there are
+// transactions that will node fit in the given square size, then they are
+// discarded. This is reflected in the returned block data. Note: pointers to
+// block data are only used to avoid dereferening, not because we need the block
+// data to be mutable.
+func SplitShares(txConf client.TxConfig, squareSize uint64, data *core.Data) ([][]byte, *core.Data) {
+	var (
+		processedTxs [][]byte
+		messages     core.Messages
+	)
+	sqwr := newShareSplitter(txConf, squareSize, data)
+
+	for _, rawTx := range data.Txs {
+		// decode the Tx
+		tx, err := txConf.TxDecoder()(rawTx)
+		if err != nil {
+			continue
+		}
+
+		authTx, ok := tx.(signing.Tx)
+		if !ok {
+			continue
+		}
+
+		// write the tx to the square if it normal
+		if !hasWirePayForData(authTx) {
+			success, err := sqwr.writeTx(rawTx)
+			if err != nil {
+				continue
+			}
+			if !success {
+				// the square is full
+				break
+			}
+			processedTxs = append(processedTxs, rawTx)
+			continue
+		}
+
+		// only support malleated transactions that contain a single sdk.Msg
+		if len(authTx.GetMsgs()) != 1 {
+			continue
+		}
+
+		msg := authTx.GetMsgs()[0]
+		wireMsg, ok := msg.(*types.MsgWirePayForData)
+		if !ok {
+			continue
+		}
+
+		// run basic validation on the transaction (which include the wireMsg
+		// above)
+		err = authTx.ValidateBasic()
+		if err != nil {
+			continue
+		}
+
+		// attempt to malleate and write the resulting tx + msg to the square
+		parentHash := sha256.Sum256(rawTx)
+		success, malTx, message, err := sqwr.writeMalleatedTx(parentHash[:], authTx, wireMsg)
+		if err != nil {
+			continue
+		}
+		if !success {
+			// the square is full, but we will attempt to continue to fill the
+			// block until there are no tx left or no room for txs. While there
+			// was not room for this particular tx + msg, there might be room
+			// for other txs or even other smaller messages
+			continue
+		}
+		processedTxs = append(processedTxs, malTx)
+		messages.MessagesList = append(messages.MessagesList, message)
+	}
+
+	sort.Slice(messages.MessagesList, func(i, j int) bool {
+		return bytes.Compare(messages.MessagesList[i].NamespaceId, messages.MessagesList[j].NamespaceId) < 0
+	})
+
+	return sqwr.export(), &core.Data{
+		Txs:      processedTxs,
+		Messages: messages,
+		Evidence: data.Evidence,
+	}
+}
+
+// shareSplitter write a data square using provided block data. It also ensures
+// that message and their corresponding txs get written to the square
+// atomically.
+type shareSplitter struct {
+	txWriter  *coretypes.ContiguousShareWriter
+	msgWriter *coretypes.MessageShareWriter
+
+	// we aren't doing anything with evidence or isrs yet, so we don't need to
+	// lazily generate those shares lazily
+	evdShares [][]byte
```

Comment:
Also, what do you mean by we are not doing anything with evidence?

Reply:
https://github.com/celestiaorg/celestia-app/pull/224/commits/9dc8d123484dce907781f97a855657b239c26327